main/src/ca/uwaterloo/flix/language/phase/EnumGen.scala

```diff
@@ -0,0 +1,650 @@
+    * we generate:
+    *
+    * public int value;
+    *
+    * but for the case `Ok[List[Int32]]` we generate:
+    *
+    * public Object value;
+    *
+    * Classes generated at this step implements the interface corresponding the symbol of the enum case and they include
+    * implementations of following methods: `getTag()`, `getValue()`, `getBoxedValue()`,`toString()`, `hashCode()` and
+    * `equals(Object)`.
+    * `getTag()` is the function which returns the value of the `tag` field. `getValue()` returns the value of `value` field.
+    * `getBoxedValue()` returns the `value` field but the result is boxed inside an object. As an example, `getValue()` and
+    * `getBoxedValue()` of the class representing `Ok[Int32]` is as follows:
+    *
+    * public final int getValue() {
+    *   return this.value;
+    * }
+    *
+    * public final Object getBoxedValue() {
+    *   return new Integer(this.value);
+    * }
+    *
+    * `toString()` returns a representation of the enum case with `CaseName(fieldRepresentation)` as its format. For example
+    * for `Some[Int]` we generate the following method:
+    *
+    * public String toString() {
+    *   return "Some(".concat(String.valueOf(this.value).concat(")"));
+    * }
+    *
+    * `hashCode()` returns a hashCode corresponding to this enum case. hashCode is the the hash of a list containing namespace
+    * of the enum symbol, enum name and case name multiplied by 7 added to the hashCode of the field of the enum multiplied by 11.
+    * For example, for enum `Some[Bool]` we generate the following method:
+    *
+    * public int hashCode() {
+    *   return 7 * -1721515276 + 11 * this.value;
+    * }
+    *
+    * and for `Some[List[Int]]` we generate the following method:
+    *
+    * public int hashCode() {
+    *   return 7 * -1721515276 + 11 * this.value;
+    * }
+    *
+    * Finally, for `equals(object)`, we first check that `object` is of the type of the class, then we invoke `equals()`
+    * method on the `value` field of `this` and `object`. If the field is primitive then we just use `==` sign to compare
+    * fields. For example for `Result[Int32]` we generate:
+    *
+    * public boolean equals(Object var1) {
+    *   return var1 instanceof Ok && this.value == ((Ok)var1).value;
+    * }
+    *
+    * and for `Err[List[Int32]]` we generate:
+    *
+    * public boolean equals(Object var1) {
+    *   return var1 instanceof Err?this.value.equals(((Err)var1).value):false;
+    * }
+    */
+  def run(root: ExecutableAst.Root)(implicit flix: Flix): Validation[ExecutableAst.Root, CompilationError] = {
+    implicit val _ = flix.genSym
+
+    val t = System.nanoTime()
+
+    if (flix.options.evaluation == Evaluation.Interpreted) {
+      return root.toSuccess
+    }
+
+    // 1. Transform non-functions and generate interface names.
+    val constantsList: List[Definition.Constant] = root.definitions.values.map { f =>
```

Comment:
!!!: You are not supposed to change the AST. What is the reason to change the type?

(NB: The types in Code Generator have been a source of bugs/confusion so it is entirely possible that what you are doing makes good sense, but in that case the transformation should probably happen earlier in the compiler),